src/ui/common/src/components/pages/metric/id/index.tsx

```diff
@@ -50,109 +42,68 @@ const MetricDetailsPage: React.FC<MetricDetailsPageProps> = ({
   Layout = DefaultLayout,
 }) => {
   const dispatch: AppDispatch = useDispatch();
-  const navigate = useNavigate();
   const { workflowId, workflowDagResultId, metricOperatorId } = useParams();
 
   const [inputsExpanded, setInputsExpanded] = useState<boolean>(true);
   const [outputsExpanded, setOutputsExpanded] = useState<boolean>(true);
 
-  const workflow = useSelector((state: RootState) => state.workflowReducer);
-  const workflowDag = workflow.dagResults.find((currentDag) => {
-    return currentDag.id === workflowDagResultId;
-  });
-
-  const workflowDagId = workflowDag?.workflow_dag_id;
-  const operator = (workflow.dags[workflowDagId]?.operators ?? {})[
-    metricOperatorId
-  ];
-
-  const metricResult = workflow.operatorResults[metricOperatorId];
+  const workflowDagResultWithLoadingStatus = useSelector(
+    (state: RootState) =>
+      state.workflowDagResultsReducer.results[workflowDagResultId]
+  );
+  const operator = (workflowDagResultWithLoadingStatus?.result?.operators ??
+    {})[metricOperatorId];
+  const artifactId = operator?.outputs[0];
+  const artifactHistoryWithLoadingStatus = !!artifactId
+    ? useSelector(
+        (state: RootState) => state.artifactResultsReducer.artifacts[artifactId]
+      )
+    : undefined;
 
   useEffect(() => {
     // TODO: Update this to contain the name of the operator
     document.title = 'Metric Details | Aqueduct';
-    dispatch(handleGetWorkflow({ apiKey: user.apiKey, workflowId }));
-  }, []);
 
-  useEffect(() => {
-    if (workflow.dagResults.length > 0) {
+    // Load workflow dag result if it's not cached
+    if (
+      !workflowDagResultWithLoadingStatus ||
+      isInitial(workflowDagResultWithLoadingStatus.status)
+    ) {
       dispatch(
-        handleGetOperatorResults({
+        handleGetWorkflowDagResult({
           apiKey: user.apiKey,
+          workflowId,
           workflowDagResultId,
-          operatorId: metricOperatorId,
         })
       );
     }
-  }, [workflow.dagResults]);
+  }, []);
 
   useEffect(() => {
-    if (!metricResult || !metricResult.result) {
-      return;
+    // Load artifact history once workflow dag results finished loading
+    // and the result is not cached
```

Comment:
What does "and the result is not cached" mean? Is it referring to the fact we query the api each time workflowDagResultWithLoadingStatus or artifactId is changed?

Reply:
redux store can also serve as a cache. If `artifactHistoryWithLoadingStatus` is not `undefined`, that means there's an entry in entry in `artifactResultsReducer` for the given `artifactId`. In this case we don't have to fetch the result again. Otherwise the result is missing and we dispatch the handler.

And yes, this caching help us to not call this API every time the wfDagResult is changed